feat: Union types coercion

gandronchik opened the pull request:
# Which issue does this PR close?

[Union columns with different types](https://github.com/apache/arrow-datafusion/issues/3467)

closes https://github.com/apache/arrow-datafusion/issues/3467

alamb reviewed (commented):
Thanks @gandronchik  -- I think this PR is looking really good.

The only thing I think is needed prior to merge is 2 more tests. ❤️ Thank you!

gandronchik commented:
> Thanks @gandronchik -- I think this PR is looking really good.
> 
> The only thing I think is needed prior to merge is 2 more tests. ❤️ Thank you!

Hi @alamb. Thank you for review. I added more tests and also fixed few bugs 🙏

alamb reviewed (approved):
This looks great -- thank you @gandronchik . I went over the tests carefully and 👍
